### PR TITLE
Add superobbing for AHI

### PR DESCRIPTION
**Pull request by YaliWu0219**
TYPE: new feature

KEYWORDS: superobbing, cloud detection, cloud fraction, AHI 

SOURCE: Yali Wu (NCAR/MMM), Jonathan Guerrette (NCAR/MMM).

DESCRIPTION OF CHANGES: This PR adds a new function of superobbing AHI radiances and retrieving cloud fraction based on AHI cloud detection results. The default namelist parameter `ahi_superob_halfwidth` is set to zero, which means no superobbing will be applied.  

ISSUE: no issue

LIST OF MODIFIED FILES: 
M       Registry/registry.var
M       var/da/da_define_structures/da_define_structures.f90
M       var/da/da_monitor/da_rad_diags.f90
M       var/da/da_radiance/da_allocate_rad_iv.inc
M       var/da/da_radiance/da_deallocate_radiance.inc
M       var/da/da_radiance/da_initialize_rad_iv.inc
M       var/da/da_radiance/da_qc_ahi.inc
M       var/da/da_radiance/da_radiance.f90
M       var/da/da_radiance/da_radiance1.f90
M       var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc
M       var/da/da_radiance/da_write_iv_rad_ascii.inc
M       var/da/da_radiance/da_write_oa_rad_ascii.inc

TESTS CONDUCTED:
This capability has been tested in generating one-month AHI obs files for MPAS-JEDI DA purpose. WRFDA regression tests passed.

A simple comparison between using thinning and superobbing:
image
And a figure of cloud fraction product from a typhoon event:
image

### CODE REVIEW

**Comment by davegill** [comment]
@YaliWu0219 
I am seeing a failure for one of the tests.

It is possible that we fixed this problem in release-v4.2, since I do not see this problem with the release-v4.2 tests that have been conducted for the last few weeks.

In a separate (throw away) branch, merge your mods into release-v4.2, and then submit that _test_ branch as a PR. If all of the regression tests PASS in that _test_ PR, post that information in _this_ PR commit message.

**Comment by davegill** [comment]
@YaliWu0219 
Here are the jenkins test results:
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 6735270dd3af552d8e2bb067466dc82f3f2ee701, requested by: YaliWu0219 for PR: https://github.com/wrf-model/WRF/pull/1173. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 18           18
    Number of Builds       : 46           46
    Number of Simulations  : 164           164         3
    Number of Comparisons  : 104           102         0

    Failed Simulations are: 
    output_1:9 = STATUS test_001s em em_real 32 rap
output_1:9 = STATUS test_001o em em_real 33 rap
output_1:9 = STATUS test_001m em em_real 34 rap
    Which comparisons are not bit-for-bit: 
    None
```

When I look at the output for the "rap" namelist:
```
Backtrace for this error:
#0  0x14cb27463dfd in ???
#1  0x14cb27463013 in ???
#2  0x14cb2695b3af in ???
#3  0x2ada3d8 in __module_sf_mynn_MOD_psim_stable
        at /wrf/WRF/phys/module_sf_mynn.f90:2050
#4  0x2adab5c in __module_sf_mynn_MOD_zolri2
        at /wrf/WRF/phys/module_sf_mynn.f90:1965
#5  0x2adae16 in __module_sf_mynn_MOD_zolri
        at /wrf/WRF/phys/module_sf_mynn.f90:1912
```

**Comment by YaliWu0219** [comment]
@davegill Thank you for showing me this information. It seems that there's no response when I click `Details`. I will try that on  release-v4.2.

**Comment by davegill** [comment]
@YaliWu0219 
Yes, that "click details" button does not work. We are not sure how to get rid of it!

If you have questions on how to set up the temporary test branch, please send email to gill@ucar.edu

File: var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc [hunk 1]
```diff
@@ -510,17 +516,74 @@ end if
...
+               if (ahi_superob_halfwidth .gt.0 .and. nkeep .gt. 0) then
+                  tb = temp1 / real(nkeep,8)
+               else
+                  ! Extract single pixel BT and radiance value for this channel
+                  tb = tbb(ilongitude,ilatitude,k)
+               end if
+
+               if (tb > 0.0)   p % tb_inv(k) = tb
+
+             end do
```
Comment (liujake) [inline var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc:548]: this portion of code seems to do the superobing, but I do not see how superobed tb is used in late code.

File: var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc [hunk 2]
```diff
@@ -510,17 +516,74 @@ end if
...
+                  iysup = ilatitude + jsup-1-ahi_superob_halfwidth
+                  ixsup = ilongitude + isup-1-ahi_superob_halfwidth
+                  if (ixsup < 1 .or. ixsup > nlongitude .or. iysup < 1 .or. iysup > nlatitude ) then
+                     cycle
+                  end if
+                  if ( .not. allocated(p % superob(isup,jsup) % tb_obs ) ) then
+                     allocate ( p % superob(isup,jsup) % tb_obs  (1:nchan,1) )
+                     allocate ( p % superob(isup,jsup) % cld_qc(1) )
+                  end if
+                  p % superob(isup,jsup) % tb_obs(1:nchan,1)      = tbb( ixsup, iysup, 1:nchan )
```
Comment (liujake) [inline var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc:571]: Here seems to simply assign raw tbb data into superob array. Previously superobed data is not used.

Comment (YaliWu0219) [reply var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc:571]: `p % superob(isup,jsup) % tb_obs(1:nchan,1)` is for cloud detection use only, where we should retain the raw tbb value.

File: var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc (hunk 1 above)
Comment (YaliWu0219) [reply var/da/da_radiance/da_read_obs_netcdf4ahi_jaxa.inc:548]: Superobbed value is used as `p % tb_inv(k) = tb`.

**Comment by weiwangncar** [comment]
@YaliWu0219 At the moment, the WRF model only computes cloud fraction as a diagnostics for radiation. Do you expect more from the model?

**Comment by liujake** [comment]
@weiwangncar This is nothing to do with model cloud fraction.

**Comment by liujake** [comment]
> @YaliWu0219
> Here are the jenkins test results:
> 
> ```
> Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 6735270dd3af552d8e2bb067466dc82f3f2ee701, requested by: YaliWu0219 for PR: https://github.com/wrf-model/WRF/pull/1173. For any query please send e-mail to David Gill.
> 
>     Test Type              | Expected  | Received |  Failed
>     = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
>     Number of Tests        : 18           18
>     Number of Builds       : 46           46
>     Number of Simulations  : 164           164         3
>     Number of Comparisons  : 104           102         0
> 
>     Failed Simulations are: 
>     output_1:9 = STATUS test_001s em em_real 32 rap
> output_1:9 = STATUS test_001o em em_real 33 rap
> output_1:9 = STATUS test_001m em em_real 34 rap
>     Which comparisons are not bit-for-bit: 
>     None
> ```
> 
> When I look at the output for the "rap" namelist:
> 
> ```
> Backtrace for this error:
> #0  0x14cb27463dfd in ???
> #1  0x14cb27463013 in ???
> #2  0x14cb2695b3af in ???
> #3  0x2ada3d8 in __module_sf_mynn_MOD_psim_stable
>         at /wrf/WRF/phys/module_sf_mynn.f90:2050
> #4  0x2adab5c in __module_sf_mynn_MOD_zolri2
>         at /wrf/WRF/phys/module_sf_mynn.f90:1965
> #5  0x2adae16 in __module_sf_mynn_MOD_zolri
>         at /wrf/WRF/phys/module_sf_mynn.f90:1912
> ```

@davegill Is this issue in the 'develop' branch (not in v4.2 branch) already fixed?

**Comment by davegill** [comment]
@liujake 
> Is this issue in the 'develop' branch (not in v4.2 branch) already fixed?

Jake,
Yes.

**Pull request merged**